File: src/utils/paths.ts
import * as path from 'path';
import * as fs from 'fs';

const CLRUN_DIR = '.clrun';

/**
 * Resolve the project root by walking up from cwd looking for common
 * project indicators, falling back to cwd itself.
 */
export function resolveProjectRoot(): string {
  let dir = process.cwd();
  const root = path.parse(dir).root;

  while (dir !== root) {
    const indicators = ['package.json', '.git', 'Cargo.toml', 'go.mod', 'pyproject.toml', 'Makefile'];
    for (const indicator of indicators) {
      if (fs.existsSync(path.join(dir, indicator))) {
        return dir;
      }
    }
    // Also check if .clrun already exists
    if (fs.existsSync(path.join(dir, CLRUN_DIR))) {
      return dir;
    }
    dir = path.dirname(dir);
  }

  return process.cwd();
}

/**
 * Get all .clrun paths for the project.
 */
export function getClrunPaths(projectRoot?: string) {
  const root = projectRoot ?? resolveProjectRoot();
  const clrunDir = path.join(root, CLRUN_DIR);

  return {
    root: clrunDir,
    runtimeLock: path.join(clrunDir, 'runtime.lock'),
    runtimePid: path.join(clrunDir, 'runtime.pid'),
    runtimeJson: path.join(clrunDir, 'runtime.json'),
    sessionsDir: path.join(clrunDir, 'sessions'),
    queuesDir: path.join(clrunDir, 'queues'),
    buffersDir: path.join(clrunDir, 'buffers'),
    ledgerDir: path.join(clrunDir, 'ledger'),
    eventsLog: path.join(clrunDir, 'ledger', 'events.log'),
    skillsDir: path.join(clrunDir, 'skills'),
  } as const;
}

/**
 * Ensure all .clrun directories exist.
 */
export function ensureClrunDirs(projectRoot?: string): void {
  const paths = getClrunPaths(projectRoot);
  const dirs = [
    paths.root,
    paths.sessionsDir,
    paths.queuesDir,
    paths.buffersDir,
    paths.ledgerDir,
    paths.skillsDir,
  ];

  for (const dir of dirs) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

/**
 * Get session file path for a terminal.
 */
export function sessionPath(terminalId: string, projectRoot?: string): string {
  const paths = getClrunPaths(projectRoot);
  return path.join(paths.sessionsDir, `${terminalId}.json`);
}

/**
 * Get queue file path for a terminal.
 */
export function queuePath(terminalId: string, projectRoot?: string): string {
  const paths = getClrunPaths(projectRoot);
  return path.join(paths.queuesDir, `${terminalId}.json`);
}

/**
 * Get buffer file path for a terminal.
 */
export function bufferPath(terminalId: string, projectRoot?: string): string {
  const paths = getClrunPaths(projectRoot);
  return path.join(paths.buffersDir, `${terminalId}.log`);
}

File: python/clrun/utils/paths.py
"""Project root resolution and .clrun path management."""

from __future__ import annotations

import os
from dataclasses import dataclass

CLRUN_DIR = ".clrun"

INDICATORS = [
    "package.json",
    ".git",
    "Cargo.toml",
    "go.mod",
    "pyproject.toml",
    "Makefile",
]


def resolve_project_root() -> str:
    """Walk up from cwd looking for project indicators, fall back to cwd."""
    d = os.getcwd()
    root = os.path.abspath(os.sep)

    while d != root:
        for indicator in INDICATORS:
            if os.path.exists(os.path.join(d, indicator)):
                return d
        if os.path.exists(os.path.join(d, CLRUN_DIR)):
            return d
        d = os.path.dirname(d)

    return os.getcwd()


@dataclass(frozen=True)
class ClrunPaths:
    root: str
    runtime_lock: str
    runtime_pid: str
    runtime_json: str
    sessions_dir: str
    queues_dir: str
    buffers_dir: str
    ledger_dir: str
    events_log: str
    skills_dir: str


def get_clrun_paths(project_root: str | None = None) -> ClrunPaths:
    pr = project_root or resolve_project_root()
    cr = os.path.join(pr, CLRUN_DIR)
    return ClrunPaths(
        root=cr,
        runtime_lock=os.path.join(cr, "runtime.lock"),
        runtime_pid=os.path.join(cr, "runtime.pid"),
        runtime_json=os.path.join(cr, "runtime.json"),
        sessions_dir=os.path.join(cr, "sessions"),
        queues_dir=os.path.join(cr, "queues"),
        buffers_dir=os.path.join(cr, "buffers"),
        ledger_dir=os.path.join(cr, "ledger"),
        events_log=os.path.join(cr, "ledger", "events.log"),
        skills_dir=os.path.join(cr, "skills"),
    )


def ensure_clrun_dirs(project_root: str | None = None) -> None:
    paths = get_clrun_paths(project_root)
    for d in [
        paths.root,
        paths.sessions_dir,
        paths.queues_dir,
        paths.buffers_dir,
        paths.ledger_dir,
        paths.skills_dir,
    ]:
        os.makedirs(d, exist_ok=True)


def session_path(terminal_id: str, project_root: str | None = None) -> str:
    return os.path.join(get_clrun_paths(project_root).sessions_dir, f"{terminal_id}.json")


def queue_path(terminal_id: str, project_root: str | None = None) -> str:
    return os.path.join(get_clrun_paths(project_root).queues_dir, f"{terminal_id}.json")


def buffer_path(terminal_id: str, project_root: str | None = None) -> str:
    return os.path.join(get_clrun_paths(project_root).buffers_dir, f"{terminal_id}.log")
